apps/settings/js/panels/simpin_dialog/simpin_dialog.js

```diff
@@ -0,0 +1,602 @@
+      }, (error) => {
+        var needToCloseDialog = this._handleCardLockError({
+          lockType: options.lockType,
+          retryCount: error.retryCount
+        });
+
+        if (needToCloseDialog) {
+          return true;
+        } else {
+          return Promise.reject();
+        }
+      });
+    },
+
+    /**
+     * Lock pin
+     *
+     * @memberOf SimPinDialog
+     * @param {Boolean} enabled
+     * @access private
+     * @return {Promise}
+     */
+    _enableLock: function(enabled) {
+      var pin = this._elements.pinInput.value;
+      if (pin === '') {
+        return Promise.reject();
+      }
+      return this._setCardLock({
+        lockType: 'pin',
+        pin: pin,
+        enabled: enabled
+      });
+    },
+
+    /**
+     * Enable FDN on Simcards and related Call functions under its own panel.
+     *
+     * @memberOf SimPinDialog
+     * @param {Boolean} enabled
+     * @access private
+     * @return {Promise}
+     */
+    _enableFdn: function(enabled) {
+      var pin = this._elements.pinInput.value;
+      if (pin === '') {
+        return Promise.reject();
+      }
+      return this._setCardLock({
+        lockType: 'fdn',
+        pin2: pin,
+        enabled: enabled
+      });
+    },
+
+    /**
+     * Change the pre-defined codes in the simcard.
+     *
+     * @memberOf SimPinDialog
+     * @param {String} lockType
+     * @access private
+     * @return {Promise}
+     */
+    _changePin: function(lockType) {
+      // lockType = `pin' or `pin2'
+      lockType = lockType || 'pin';
+      var pin = this._elements.pinInput.value;
+      var newPin = this._elements.newPinInput.value;
+      var confirmPin = this._elements.confirmPinInput.value;
+
+      if (pin === '' || newPin === '' || confirmPin === '') {
+        return Promise.reject();
+      }
+      
+      if (newPin !== confirmPin) {
+        this._showMessage('newPinErrorMsg');
+        this._elements.newPinInput.value = '';
+        this._elements.confirmPinInput.value = '';
+        return Promise.reject();
+      }
+
+      return this._setCardLock({
+        lockType: lockType,
+        pin: pin,
+        newPin: newPin
+      });
+    },
+
+    /**
+     * This is an internal function to pass all arguments to icc and waiting
+     * for returned results.
+     *
+     * @memberOf SimPinDialog
+     * @param {Object} options
+     * @param {String} options.lockType
+     * @param {String} options.pin
+     * @param {String} options.newPin
+     * @param {String} options.pin2
+     * @param {Boolean} options.enabled
+     * @access private
+     * @return {Promise}
+     */
+    _setCardLock: function(options) {
+      return this._icc.setCardLock(options).then(() => {
+        return true; 
+      }, (error) => {
+        var needToCloseDialog = this._handleCardLockError({
+          lockType: options.lockType,
+          retryCount: error.retryCount
+        });
+        if (needToCloseDialog) {
+          return true;
```

Comment:
We can simply return `Promise.rejcet` to prevent the dialog from being closed but do nothing to make it closed.